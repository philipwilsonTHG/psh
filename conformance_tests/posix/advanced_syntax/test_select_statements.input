# Test select statement functionality
# PSH supports select statements for interactive menus (v0.34.0 feature)

# Note: Select statements are primarily interactive and require user input
# This test focuses on the structure and setup of select statements

echo "Testing select statement functionality:"

# Test basic select statement structure
echo "Testing basic select statement setup:"

# Simulate select behavior with predefined choices
echo "Basic select menu structure:"
echo "1) apple"
echo "2) banana" 
echo "3) cherry"
echo "Select simulation: choice 2 would set REPLY=2 and var=banana"

# Test select with variable list
echo "Testing select with variable list:"
fruits="apple banana cherry orange"
echo "Select menu for fruits:"
count=1
for fruit in $fruits; do
    echo "$count) $fruit"
    count=$((count + 1))
done
echo "Fruits select menu created"

# Test select with command substitution
echo "Testing select with command substitution:"
echo "Files in current directory:"
count=1
for file in file1.txt file2.txt file3.log; do
    echo "$count) $file"
    count=$((count + 1))
done
echo "File select menu created"

# Test select with array-like behavior
echo "Testing select with multiple items:"
set -- red green blue yellow
echo "Color menu:"
count=1
for color in "$@"; do
    echo "$count) $color"
    count=$((count + 1))
done
echo "Color select menu created"

# Test PS3 prompt variable
echo "Testing PS3 prompt variable:"
PS3="Choose an option: "
echo "PS3 set to: '$PS3'"

PS3="Select> "
echo "PS3 changed to: '$PS3'"

# Test select with complex menu items
echo "Testing complex menu items:"
echo "Menu with descriptions:"
echo "1) start - Start the service"
echo "2) stop - Stop the service"
echo "3) restart - Restart the service"
echo "4) status - Check service status"
echo "5) quit - Exit menu"
echo "Complex menu created"

# Test select menu formatting
echo "Testing menu formatting:"

# Multi-column simulation
items="item1 item2 item3 item4 item5 item6 item7 item8"
echo "Multi-column menu simulation:"
count=1
cols=3
col=1
for item in $items; do
    printf "%2d) %-10s" $count "$item"
    if [ $col -eq $cols ]; then
        echo
        col=1
    else
        col=$((col + 1))
    fi
    count=$((count + 1))
done
echo
echo "Multi-column menu formatted"

# Test select with special characters
echo "Testing special characters in menu:"
echo "1) file-with-dashes.txt"
echo "2) file_with_underscores.txt"
echo "3) file with spaces.txt"
echo "4) file.with.dots.txt"
echo "Special character menu created"

# Test empty menu handling
echo "Testing empty menu:"
empty_list=""
if [ -z "$empty_list" ]; then
    echo "Empty list detected - menu would exit immediately"
fi

# Test single item menu
echo "Testing single item menu:"
single_item="only_choice"
echo "1) $single_item"
echo "Single item menu created"

# Test very long menu
echo "Testing long menu:"
echo "Long menu simulation (10 items):"
for i in $(seq 1 10); do
    echo "$i) Option number $i"
done
echo "Long menu created"

# Test select with break/continue concepts
echo "Testing select control flow concepts:"
echo "Select loop would support:"
echo "  - break: exit the select loop"
echo "  - continue: redisplay menu and prompt"
echo "  - Invalid choice: redisplay menu"

# Test select with nested structures
echo "Testing nested select concept:"
echo "Main menu:"
echo "1) Files"
echo "2) Directories" 
echo "3) Processes"
echo ""
echo "Submenu for Files:"
echo "1) List files"
echo "2) Create file"
echo "3) Delete file"
echo "4) Back to main menu"
echo "Nested menu structure demonstrated"

# Test REPLY variable concept
echo "Testing REPLY variable concept:"
echo "In select loop:"
echo "  REPLY contains the user's numeric choice"
echo "  Selected variable contains the chosen item"
echo "  Example: REPLY=2, fruit=banana"

# Test select with functions
menu_function() {
    local items="$*"
    echo "Function menu for: $items"
    local count=1
    for item in $items; do
        echo "$count) $item"
        count=$((count + 1))
    done
}

echo "Testing select in functions:"
menu_function red green blue

# Test select error handling concepts
echo "Testing error handling concepts:"
echo "Select handles:"
echo "  - EOF (Ctrl-D): exit select loop"
echo "  - Interrupt (Ctrl-C): exit select loop"
echo "  - Invalid input: redisplay menu"
echo "  - Empty input: redisplay menu"

# Restore original PS3
PS3="#? "
echo "PS3 restored to default: '$PS3'"

echo "Select statement test completed"