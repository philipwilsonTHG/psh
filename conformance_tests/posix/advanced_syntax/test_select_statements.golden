Testing select statement functionality:
Testing basic select statement setup:
Basic select menu structure:
1) apple
2) banana
3) cherry
Select simulation: choice 2 would set REPLY=2 and var=banana
Testing select with variable list:
Select menu for fruits:
1) fruits
Fruits select menu created
Testing select with command substitution:
Files in current directory:
1) file1.txt
2) file2.txt
3) file3.log
File select menu created
Testing select with multiple items:
Color menu:
1) red
2) green
3) blue
4) yellow
Color select menu created
Testing PS3 prompt variable:
PS3 set to: 'Choose an option: '
PS3 changed to: 'Select> '
Testing complex menu items:
Menu with descriptions:
1) start - Start the service
2) stop - Stop the service
3) restart - Restart the service
4) status - Check service status
5) quit - Exit menu
Complex menu created
Testing menu formatting:
Multi-column menu simulation:
 1) items
Multi-column menu formatted
Testing special characters in menu:
1) file-with-dashes.txt
2) file_with_underscores.txt
3) file with spaces.txt
4) file.with.dots.txt
Special character menu created
Testing empty menu:
Empty list detected - menu would exit immediately
Testing single item menu:
1) only_choice
Single item menu created
Testing long menu:
Long menu simulation (10 items):
1) Option number 1
2) Option number 2
3) Option number 3
4) Option number 4
5) Option number 5
6) Option number 6
7) Option number 7
8) Option number 8
9) Option number 9
10) Option number 10
Long menu created
Testing select control flow concepts:
Select loop would support:
  - break: exit the select loop
  - continue: redisplay menu and prompt
  - Invalid choice: redisplay menu
Testing nested select concept:
Main menu:
1) Files
2) Directories
3) Processes

Submenu for Files:
1) List files
2) Create file
3) Delete file
4) Back to main menu
Nested menu structure demonstrated
Testing REPLY variable concept:
In select loop:
  REPLY contains the user's numeric choice
  Selected variable contains the chosen item
  Example: REPLY=2, fruit=banana
Testing select in functions:
Function menu for: red green blue
1) items
Testing error handling concepts:
Select handles:
  - EOF (Ctrl-D): exit select loop
  - Interrupt (Ctrl-C): exit select loop
  - Invalid input: redisplay menu
  - Empty input: redisplay menu
PS3 restored to default: '#? '
Select statement test completed
